### PR TITLE
Feature/niad 904 topic category lists in encounter mapping

### PR DESCRIPTION
**Pull request by msolecka**

### CODE REVIEW

File: service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/utils/ResourceExtractor.java [hunk 1]
```diff
@@ -2,19 +2,63 @@
...
+    }
+
+    private static boolean hasId(IIdType reference, Resource resource) {
+        return resource.hasIdElement() && StringUtils.equals(resource.getIdElement().getIdPart(), reference.getIdPart());
+    }
+
+    private static boolean hasEncounterReference(IIdType reference, ListResource resource) {
+        return resource.hasEncounter()
+            && resource.getEncounter().hasReferenceElement()
+            && StringUtils.equals(resource.getEncounter().getReferenceElement().getIdPart(), reference.getIdPart());
```
Comment (mmackiewicz1) [inline service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/utils/ResourceExtractor.java:58]: Could we just use "equals" String object method?

File: service/src/test/java/uk/nhs/adaptors/gp2gp/ehr/mapper/EncounterComponentsMapperTest.java [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+
+    private Optional<Encounter> extractEncounter(Bundle bundle) {
+        return bundle.getEntry().stream()
+            .map(Bundle.BundleEntryComponent::getResource)
+            .filter(resource -> ResourceType.Encounter.equals(resource.getResourceType()))
+            .map(resource -> (Encounter) resource)
+            .findFirst();
+    }
+
+    private static Stream<Arguments> testData() {
```
Comment (mmackiewicz1) [inline service/src/test/java/uk/nhs/adaptors/gp2gp/ehr/mapper/EncounterComponentsMapperTest.java:112]: Is there a point of making a parametrized test for only one set of test data?

Comment (msolecka) [reply service/src/test/java/uk/nhs/adaptors/gp2gp/ehr/mapper/EncounterComponentsMapperTest.java:112]: right, I should change that.

File: service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/utils/ResourceExtractor.java (hunk 1 above)
Comment (msolecka) [reply service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/utils/ResourceExtractor.java:58]: This utils also compares nulls, and due to that there is no need to check if in this case idPart is not null before calling equal on it, so preventing NullPointerException. I think it's better to use those string utils to compare strings.

File: service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/EhrExtractMapper.java [hunk 3]
```diff
@@ -67,13 +57,17 @@ private EhrExtractTemplateParameters prepareEhrFhirExtractParamsFromFhirBundle(
...
 
     private Optional<Patient> extractPatientFromBundle(Bundle bundle) {
-        return bundle.getEntry()
-            .stream()
-            .filter(entry -> !entry.isEmpty())
-            .map(Bundle.BundleEntryComponent::getResource)
-            .filter(resource -> resource.getResourceType() == ResourceType.Patient)
-            .map(resource -> (Patient) resource)
-            .findFirst();
+        if (bundle != null && bundle.hasEntry()) {
```
Comment (mattd-kainos) [inline service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/EhrExtractMapper.java:60]: I'm confused as to why this method exists at all. The returned patient is never used anywhere, we just throw an exception if its missing.

**Review by mmackiewicz1 — DISMISSED** [review]

**Review by ElliotB-96 — DISMISSED** [review]

File: service/src/intTest/java/uk/nhs/adaptors/gp2gp/gpc/GetGpcStructuredComponentTest.java [hunk 4]
```diff
@@ -156,6 +157,7 @@ private void assertThatObjectCreated(StorageDataWrapper storageDataWrapper, EhrE
         String hl7Response = storageDataWrapper.getHl7Response();
         assertThat(hl7Response).contains(EXPECTED_PAYLOAD_TYPE);
         assertThat(hl7Response).contains(EHR_COMPOSITION_ELEMENT);
+        assertThat(hl7Response).containsIgnoringCase(COMPONENT_ELEMENT);
```
Comment (mattd-kainos) [inline service/src/intTest/java/uk/nhs/adaptors/gp2gp/gpc/GetGpcStructuredComponentTest.java:160]: why ignoring case?

File: service/src/intTest/java/uk/nhs/adaptors/gp2gp/gpc/GetGpcStructuredComponentTest.java [hunk 5]
```diff
@@ -63,7 +64,7 @@
     private DetectTranslationCompleteService detectTranslationCompleteService;
 
     @Test
-    public void When_NewStructuredTask_Expect_DatabaseUpdatedAndAddedToObjectStore() throws IOException {
+    public void When_NewStructuredTask0_Expect_DatabaseUpdatedAndAddedToObjectStore() throws IOException {
```
Comment (mattd-kainos) [inline service/src/intTest/java/uk/nhs/adaptors/gp2gp/gpc/GetGpcStructuredComponentTest.java:67]: NewStructuredTask0? Why the name change?

Comment (msolecka) [reply service/src/intTest/java/uk/nhs/adaptors/gp2gp/gpc/GetGpcStructuredComponentTest.java:67]: haven't noticed it sorry.

File: service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/EncounterComponentsMapper.java [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+    public String mapComponents(Encounter encounter) {
+        Optional<ListResource> listReferencedToEncounter =
+            messageContext.getInputBundleHolder().getListReferencedToEncounter(encounter.getIdElement(), CONSULTATION_LIST_CODE);
+
+        return listReferencedToEncounter
+            .map(this::mapItemsTpComponents)
+            .orElse(StringUtils.EMPTY);
+    }
+
+    private String mapItemsTpComponents(ListResource listReferencedToEncounter) {
```
Comment (mattd-kainos) [inline service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/EncounterComponentsMapper.java:67]: typo ("Tp") and different method name suggestion to not be confused with the next method
```suggestion
    private String mapListResourceToComponents(ListResource listReferencedToEncounter) {
```

File: service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/EhrExtractMapper.java (hunk 3 above)
Comment (msolecka) [reply service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/EhrExtractMapper.java]: The verification of patient in the bundle was removed due to misunderstanding with creating it at the beginning.

**Review by mmackiewicz1 — APPROVED** [review]

**Review by mattd-kainos — APPROVED** [review]

**Pull request merged**